### ISSUE TITLE
false-negative in google_checks.xml for not being able to detect requirement of K & R style for FINALLY

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#15664]
I have read documentation: https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style, https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/dbe82ad_2024161500/google_style.html#a4.1.3
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

From: [4.1.3 Empty blocks: may be concise](https://google.github.io/styleguide/javaguide.html#s4.1.3-braces-empty-blocks):

> An empty block or block-like construct may be in K & R style (as described in [Section 4.1.2](https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style)). Alternatively, it may be closed immediately after it is opened, with no characters or line break in between ({}), unless it is part of a multi-block statement (one that directly contains multiple blocks: if/else or try/catch/finally).

Rule states that if an empty block is part of multi-block statement then it should follow K & R style blocks.

```
$ cat Test.java 
/** some. */
public class Test {
  void foo() throws Exception {

    try (MyResource r = new MyResource()) {}
    try (MyResource r = new MyResource()) {

    } finally { after = true; }
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) {

    } finally { boolean after = true; }

    try (MyResource r = new MyResource()) {

    } finally {
      boolean after = true;
    }
  }

  /** some. */
  public class MyResource implements AutoCloseable {
    /** some. */
    @Override
    public void close() throws Exception {
      System.out.println("Closed MyResource");
    }
  }
}

$ java -jar checkstyle-10.19.0-SNAPSHOT-all.jar -c /google_checks.xml Test.java 
Starting audit...
[WARN] /var/tmp/Test.java:8:15: '{' at column 15 should have line break after. [LeftCurlyEol]
[WARN] /var/tmp/Test.java:13:15: '{' at column 15 should have line break after. [LeftCurlyEol]
Audit done.
```

**Expected**: violations for FINALLY blocks on `}` on line 8 and 13 as it should be on new line.

----------

```java
$ cat EmptyBlocks.java 
/** some. */
public class EmptyBlocks {
  void foo() throws Exception {
    int a = 90;

    if (a == 1) {
    } else {} // false-negative

    if (a == 1) {
    } else { } // false-negative

    try (MyResource r = new MyResource()) { }
    try (MyResource r = new MyResource()) {}
    try (MyResource r = new MyResource()) {} catch (Exception expected) {} // false-negative
    try (MyResource r = new MyResource()) {} catch (Exception expected) { } // false-negative
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) {} // false-negative
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) { } // false-negative
    try (MyResource r = new MyResource()) {;}
  }

  /** some. */
  public class MyResource implements AutoCloseable {
    /** some. */
    @Override
    public void close() throws Exception {
      System.out.println("Closed MyResource");
    }
  }
}
```

**Output:**

```
$ java -jar checkstyle-10.18.1-all.jar -c google_checks.xml EmptyBlocks.java
Starting audit...
[WARN] C:\checkstyle testing\EmptyBlocks.java:10: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:12: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:15: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:21: Empty blocks should have no spaces. Empty blocks                                    may only be represented[WARN] C:\checkstyle testing\EmptyBlocks.java:15: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:21: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:21: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:22:43: '{' at column 43 should have line break after. [LeftCurly]
 as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:22:43: '{' at column 43 should have line break after. [LeftCurly]
[WARN] C:\checkstyle testing\EmptyBlocks.java:22:43: '{' at column 43 should have line break after. [LeftCurly]
[WARN] C:\checkstyle testing\EmptyBlocks.java:22:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:22:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
 as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:22:45: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:22:45: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
```

We're missing violations for not following K&R style blocks though we're able to detect other mis-followed rules like having `{ }` in-place of `{}`, in current style guide implementation we don't have support to check for multi-block statements.

As part of this issue, we need to add support to detect multi-block statements and give violations if rule isn't being followed properly.

The solution I can think of is modifying following suppression:

https://github.com/checkstyle/checkstyle/blob/888ac426c1c0e45cbb4d9232dcf0c541a86a245d/src/main/resources/google_checks.xml#L107-L112

and check if we're in multi-block statement then we should give violation `'}' should be alone on the line` for else/catch/finally's right curly `}`

**Comment by romani** [comment]
We will deal with false negatives on crazy unrealistic code a later. 
I am even ok to drop it from summer project.
We will make decision a bit later.

**Comment by Zopsss** [comment]
> and make suppression 
> https://github.com/checkstyle/checkstyle/blob/f66c74cb841544b72719bc7cfaa33f7d82db7ab0/src/main/resources/google_checks.xml#L107-L112
> to not match is there is LITERAL_ELSE and LITERAL_CATCH and LITERAL_FINALLY

I was thinking that in suppression we're not checking for `if/try`'s KR block style, we're only checking for `else/catch/finally`'s KR style. We will need to ensure that if `else/catch/finally` are present then `if/try` should also follow KR style. But these tokens are already used by RightCurlySame so we cannot check this.

**Comment by romani** [comment]
@Zopsss 
here is how to update xpath to match what we need:
```
$ cat Test.java 
/** some. */
public class Test {
  void foo() throws Exception {
    int a = 90;

    if (a == 1) {
    } else {} // false-negative

    if (a == 1) {
    } else { } // false-negative

    try (MyResource r = new MyResource()) { }
    try (MyResource r = new MyResource()) {}
    try (MyResource r = new MyResource()) {} catch (Exception expected) {} // false-negative
    try (MyResource r = new MyResource()) {} catch (Exception expected) { } // false-negative
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) {} // false-negative
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) { } // false-negative
    try (MyResource r = new MyResource()) {;}
  }

  /** some. */
  public class MyResource implements AutoCloseable {
    /** some. */
    @Override
    public void close() throws Exception {
      System.out.println("Closed MyResource");
    }
  }
}

$ java -jar checkstyle-10.18.1-all.jar -b "//RCURLY[parent::SLIST[count(./*)=1] and (parent::*/parent::LITERAL_ELSE or parent::*/parent::LITERAL_CATCH or parent::*/parent::LITERAL_FINALLY)]" Test.java 
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   `--SLIST -> { [3:30]
        |       |--LITERAL_IF -> if [6:4]
        |       |   `--LITERAL_ELSE -> else [7:6]
        |       |       `--SLIST -> { [7:11]
        |       |           `--RCURLY -> } [7:12]
---------
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   `--SLIST -> { [3:30]
        |       |--LITERAL_IF -> if [9:4]
        |       |   `--LITERAL_ELSE -> else [10:6]
        |       |       `--SLIST -> { [10:11]
        |       |           `--RCURLY -> } [10:13]
---------
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   `--SLIST -> { [3:30]
        |       |--LITERAL_TRY -> try [12:4]
        |       |   `--LITERAL_CATCH -> catch [12:45]
        |       |       `--SLIST -> { [12:72]
        |       |           `--RCURLY -> } [12:73]
---------
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   `--SLIST -> { [3:30]
        |       |--LITERAL_TRY -> try [13:4]
        |       |   `--LITERAL_CATCH -> catch [13:45]
        |       |       `--SLIST -> { [13:72]
        |       |           `--RCURLY -> } [13:74]
---------
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   `--SLIST -> { [3:30]
        |       |--LITERAL_TRY -> try [14:4]
        |       |   `--LITERAL_CATCH -> catch [16:6]
        |       |       `--SLIST -> { [16:33]
        |       |           `--RCURLY -> } [16:34]
---------
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   `--SLIST -> { [3:30]
        |       |--LITERAL_TRY -> try [17:4]
        |       |   `--LITERAL_CATCH -> catch [19:6]
        |       |       `--SLIST -> { [19:33]
        |       |           `--RCURLY -> } [19:35]
```

I removed `or preceding-sibling::*[last()][self::LCURLY]]` , I do not remember why we need it, lets test it by diff report to see what for we added it.
Please also look at git history on when we added this `or ....` expression to see what use cases it covered.

**Comment by Zopsss** [comment]
I found the commit: https://github.com/checkstyle/checkstyle/pull/7542/files#diff-1fbb96c1b95e79400bf472812ee8646c92e4566b73ffc5edc83ecf146008bc23

From: https://github.com/checkstyle/checkstyle/pull/7542#issuecomment-586778038

without `or ....` part, following code is giving violations:

```
@interface TesterAnnotation {} //ok
```

```java
$ cat Test.java

/**
 * Javadoc comment.
 */
public class Test {
  @interface TesterAnnotation {} //ok
}
```

** original config:**
```xml
    <module name="SuppressionXpathSingleFilter">
      <!-- suppresion is required till https://github.com/checkstyle/checkstyle/issues/7541 -->
      <property name="id" value="RightCurlyAlone"/>
      <property name="query" value="//RCURLY[parent::SLIST[count(./*)=1]
                                     or preceding-sibling::*[last()][self::LCURLY]]"/>
    </module>
```
```
$ java -jar .\checkstyle-10.18.1-all.jar -c .\google_checks.xml Test.java
Starting audit...
Audit done.
```

**config without OR:**
```xml
    <module name="SuppressionXpathSingleFilter">
      <!-- suppresion is required till https://github.com/checkstyle/checkstyle/issues/7541 -->
      <property name="id" value="RightCurlyAlone"/>
      <property name="query" value="//RCURLY[parent::SLIST[count(./*)=1]]"/>
    </module>
```
```
java -jar .\checkstyle-10.18.1-all.jar -c .\google_checks.xml Test.java
Starting audit...
[WARN] C:\checkstyle testing\Test.java:6:32: '}' at column 32 should be alone on a line. [RightCurlyAlone]
Audit done.
```

We first encountered this case as false-negative in the report you generated back then in that PR: https://github.com/checkstyle/checkstyle/pull/7542#issuecomment-585217824

**Comment by Zopsss** [comment]
```java
$ cat .\KrBlocks.java
/** some. */
public class KrBlocks {
  @interface TesterAnnotation {} //ok

  void foo() throws Exception {
    int a = 90;

    if (a == 1) {
    } else {}

    if (a == 1) {
    } else { }

    try (MyResource r = new MyResource()) { }
    try (MyResource r = new MyResource()) {}
    try (MyResource r = new MyResource()) {} catch (Exception expected) {}
    try (MyResource r = new MyResource()) {} catch (Exception expected) { }
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) {}
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) { }
    try (MyResource r = new MyResource()) {

    } catch (Exception e) {

    } finally {}
    try (MyResource r = new MyResource()) {

    } catch (Exception e) {} finally {}
    try (MyResource r = new MyResource()) {} catch (Exception e) {} finally {}
    try (MyResource r = new MyResource()) {;}
  }

  /** some. */
  public class MyResource implements AutoCloseable {
    /** some. */
    @Override
    public void close() throws Exception {
      System.out.println("Closed MyResource");
    }
  }
}
```

**config:**
```xml
    <module name="SuppressionXpathSingleFilter">
      <!-- suppresion is required till https://github.com/checkstyle/checkstyle/issues/7541 -->
      <property name="id" value="RightCurlyAlone"/>
      <property name="query" value="//RCURLY[parent::SLIST[count(./*)=1]
                                    or preceding-sibling::*[last()][self::LCURLY]
                                    and not(parent::*/parent::LITERAL_ELSE or
                                    parent::*/parent::LITERAL_CATCH or
                                    parent::*/parent::LITERAL_FINALLY)]"/>
    </module>
```

**output:**
```
$ java -jar .\checkstyle-10.18.1-all.jar -c .\google_checks_kr.xml .\KrBlocks.java
Starting audit...
[WARN] C:\checkstyle testing\.\KrBlocks.java:12: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\KrBlocks.java:14: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\KrBlocks.java:17: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\KrBlocks.java:23: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\KrBlocks.java:26:27: Empty catch block. [EmptyCatchBlock]
[WARN] C:\checkstyle testing\.\KrBlocks.java:28:15: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:31:27: Empty catch block. [EmptyCatchBlock]
[WARN] C:\checkstyle testing\.\KrBlocks.java:31:38: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:32:66: Empty catch block. [EmptyCatchBlock]
[WARN] C:\checkstyle testing\.\KrBlocks.java:32:77: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:33:43: '{' at column 43 should have line break after. [LeftCurly]
[WARN] C:\checkstyle testing\.\KrBlocks.java:33:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:33:45: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
```

As you can see, we're still not able to detect KR blocks even after using suggested by you, but if I remove `or ....` then we're able to detect KR blocks

```xml
    <module name="SuppressionXpathSingleFilter">
      <!-- suppresion is required till https://github.com/checkstyle/checkstyle/issues/7541 -->
      <property name="id" value="RightCurlyAlone"/>
      <property name="query" value="//RCURLY[parent::SLIST[count(./*)=1]
                                    and not(parent::*/parent::LITERAL_ELSE or
                                    parent::*/parent::LITERAL_CATCH or
                                    parent::*/parent::LITERAL_FINALLY)]"/>
    </module>
```
```
java -jar .\checkstyle-10.18.1-all.jar -c .\google_checks_kr.xml .\KrBlocks.java
Starting audit...
[WARN] C:\checkstyle testing\.\KrBlocks.java:3:32: '}' at column 32 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:9:13: '}' at column 13 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:12: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\KrBlocks.java:12:14: '}' at column 14 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:14: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\KrBlocks.java:16:74: '}' at column 74 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:17: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\KrBlocks.java:17:75: '}' at column 75 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:20:35: '}' at column 35 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:23: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\KrBlocks.java:23:36: '}' at column 36 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:26:27: Empty catch block. [EmptyCatchBlock]
[WARN] C:\checkstyle testing\.\KrBlocks.java:28:5: '}' at column 5 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:28:15: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:28:16: '}' at column 16 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:31:27: Empty catch block. [EmptyCatchBlock]
[WARN] C:\checkstyle testing\.\KrBlocks.java:31:28: '}' at column 28 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:31:38: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:31:39: '}' at column 39 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:32:66: Empty catch block. [EmptyCatchBlock]
[WARN] C:\checkstyle testing\.\KrBlocks.java:32:67: '}' at column 67 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:32:77: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:32:78: '}' at column 78 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:33:43: '{' at column 43 should have line break after. [LeftCurly]
[WARN] C:\checkstyle testing\.\KrBlocks.java:33:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:33:45: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
```

we're able to detect KR block but there are some problems here.

1. At line number 3, the false negative we detected at https://github.com/checkstyle/checkstyle/issues/15664#issuecomment-2361495152 is revived again:

```
  @interface TesterAnnotation {} //ok
```

2. We're giving extra ( false-positive ) violation for `catch`'s `}`:

```java
    try (MyResource r = new MyResource()) {

    } catch (Exception e) {

    } finally {} // line 28
```

```
[WARN] C:\checkstyle testing\.\KrBlocks.java:28:5: '}' at column 5 should be alone on a line. [RightCurlyAlone]
[WARN] C:\checkstyle testing\.\KrBlocks.java:28:15: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\.\KrBlocks.java:28:16: '}' at column 16 should be alone on a line. [RightCurlyAlone]
```

ignore WhitespaceAround error Idk why that's occurring, 28:16 error is correct but 28:5 is false-positive. It is giving error for catch block's `}`

**Comment by romani** [comment]
```
$ cat Test.java 
/** some. */
public class Test {
  @interface TesterAnnotation {} //ok
}

$ java -jar checkstyle-10.18.1-all.jar -t Test.java 
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    |--MODIFIERS -> MODIFIERS [2:0]
    |   `--LITERAL_PUBLIC -> public [2:0]
    |--LITERAL_CLASS -> class [2:7]
    |--IDENT -> Test [2:13]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--LCURLY -> { [2:18]
        |--ANNOTATION_DEF -> ANNOTATION_DEF [3:2]
        |   |--MODIFIERS -> MODIFIERS [3:2]
        |   |--AT -> @ [3:2]
        |   |--LITERAL_INTERFACE -> interface [3:3]
        |   |--IDENT -> TesterAnnotation [3:13]
        |   `--OBJBLOCK -> OBJBLOCK [3:30]
        |       |--LCURLY -> { [3:30]
        |       `--RCURLY -> } [3:31]
        `--RCURLY -> } [4:0]
```
vs 
```
$ cat Test.java 
/** some. */
public class Test {
  void foo() {
    if (a) {}
    else {}
  }
}

$ java -jar checkstyle-10.18.1-all.jar -t Test.java 
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    |--MODIFIERS -> MODIFIERS [2:0]
    |   `--LITERAL_PUBLIC -> public [2:0]
    |--LITERAL_CLASS -> class [2:7]
    |--IDENT -> Test [2:13]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--LCURLY -> { [2:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   |--MODIFIERS -> MODIFIERS [3:2]
        |   |--TYPE -> TYPE [3:2]
        |   |   `--LITERAL_VOID -> void [3:2]
        |   |--IDENT -> foo [3:7]
        |   |--LPAREN -> ( [3:10]
        |   |--PARAMETERS -> PARAMETERS [3:11]
        |   |--RPAREN -> ) [3:11]
        |   `--SLIST -> { [3:13]
        |       |--LITERAL_IF -> if [4:4]
        |       |   |--LPAREN -> ( [4:7]
        |       |   |--EXPR -> EXPR [4:8]
        |       |   |   `--IDENT -> a [4:8]
        |       |   |--RPAREN -> ) [4:9]
        |       |   |--SLIST -> { [4:11]
        |       |   |   `--RCURLY -> } [4:12]
        |       |   `--LITERAL_ELSE -> else [5:4]
        |       |       `--SLIST -> { [5:9]
        |       |           `--RCURLY -> } [5:10]
        |       `--RCURLY -> } [6:2]
        `--RCURLY -> } [7:0]
```

attention to 
```
`--OBJBLOCK -> OBJBLOCK [3:30]
    |--LCURLY -> { [3:30]
    `--RCURLY -> } [3:31]
```
vs
```
`--LITERAL_ELSE -> else [5:4]
     `--SLIST -> { [5:9]
         `--RCURLY -> } [5:10]
```

**Comment by Zopsss** [comment]
> we need `or preceding-sibling::*[last()][self::LCURLY]`

yes but because of this we're not able to detect KR block style. How we should proceed with this? Should we add a message saying that KR block styles aren't supported/detected by us currently and a reference link to this issue or we should remove the `or .....` suppression, by that way we would be able to detect KR blocks but that too would contain false-positive as explained here: https://github.com/checkstyle/checkstyle/issues/15664#issuecomment-2361751694

**Comment by Zopsss** [comment]
@romani ping

**Comment by Zopsss** [comment]
@romani can you reply https://github.com/checkstyle/checkstyle/issues/15664#issuecomment-2365254452 ? Meanwhile I will try to complete other issues.

**Comment by romani** [comment]
> yes but because of this we're not able to detect KR block style.

we can detect KR style BUT there is special requirements from Google on `{}` that unfortunately we do not support by Checks gracefully, as it is edge case and not all people in world agree on this. But we do Google style, so they made decision to allow `{}`.

We do not have time in this project to make first class support of `{}`. We need to continue to use suppression.
Yeas, there will be some false-negatives (as side effect of this suppression), it is MUCH better than false-positive.
If we can narrow down suppression to minimize false-negatives, it would be awesome.

**Comment by Zopsss** [comment]
> If we can narrow down suppression to minimize false-negatives, it would be awesome.

narrowing down suppression won't help I guess. I have used both suppressions mentioned at https://github.com/checkstyle/checkstyle/issues/15664#issuecomment-2361751694 to generate diff in this PR: https://github.com/checkstyle/checkstyle/pull/15701

using `or preceding-sibling::*[last()][self::LCURLY]` does not helps us in detecting KR blocks( diff: https://github.com/checkstyle/checkstyle/pull/15701#issuecomment-2374500298 ). Removing it causes many false-positives ( you can see that in the generated diff https://github.com/checkstyle/checkstyle/pull/15701#issuecomment-2374061367 ) 

https://github.com/checkstyle/checkstyle/issues/7541 might help us to detect KR blocks, not sure with it. All maintainers needs to discuss this check, I can work on it after completing all other pending issues. This check will be out of scope of Gsoc project but solving it will help us to add support for KR blocks so I'm fine with it.

**Comment by romani** [comment]
we need to close this issue by FINALLY PR and create new more clear issue on what we are targeting as bunch of false-positives originally reported in issue are already fixed by RegexpCheck.

```
$ cat Test.java 
/** some. */
public class Test {
  void foo() throws Exception {
    int a = 90;

    if (a == 1) {
    } else {} // false-negative

    if (a == 1) {
    } else { } // false-negative

    try (MyResource r = new MyResource()) { }
    try (MyResource r = new MyResource()) {}
    try (MyResource r = new MyResource()) {} catch (Exception expected) {} // false-negative
    try (MyResource r = new MyResource()) {} catch (Exception expected) { } // false-negative
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) {} // false-negative
    try (MyResource r = new MyResource()) {

    } catch (Exception expected) { } // false-negative
    try (MyResource r = new MyResource()) {;}
  }

  /** some. */
  public class MyResource implements AutoCloseable {
    /** some. */
    @Override
    public void close() throws Exception {
      System.out.println("Closed MyResource");
    }
  }
}

$ java -jar checkstyle-10.19.0-SNAPSHOT-all.jar -c /google_checks.xml Test.java 
Starting audit...
[WARN] /var/tmp/Test.java:10: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] /var/tmp/Test.java:12: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] /var/tmp/Test.java:15: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] /var/tmp/Test.java:21: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] /var/tmp/Test.java:22:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] /var/tmp/Test.java:22:43: '{' at column 43 should have line break after. [LeftCurlyEol]
[WARN] /var/tmp/Test.java:22:45: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.

```

**Comment by romani** [comment]
Fix for "finally" is merged.
Let's create separate tickets for other tokens.
